app/js/lib/intent-parser/index.js

```diff
@@ -1,28 +1,109 @@
+
+  timeParser: Symbol('timeParser'),
+  usersParser: Symbol('usersParser'),
+  actionParser: Symbol('actionParser'),
+
+  reminderRefiner: Symbol('reminderRefiner'),
+  queryRefiner: Symbol('queryRefiner'),
 });
 
 export default class IntentParser {
   constructor() {
-    this[p.parsers] = [
-      new ReminderParser(),
-      new QueryParser(),
-    ];
+    this[p.normaliseCleaner] = new NormaliseCleaner();
+    this[p.contractionsCleaner] = new ContractionsCleaner();
+    this[p.salutationsCleaner] = new SalutationsCleaner();
+    this[p.politenessCleaner] = new PolitenessCleaner();
+
+    this[p.timeParser] = new TimeParser();
+    this[p.usersParser] = new UsersParser();
+    this[p.actionParser] = new ActionParser();
+
+    this[p.reminderRefiner] = new ReminderRefiner();
+    this[p.queryRefiner] = new QueryRefiner();
 
     window.intentParser = this;
   }
 
-  parse(phrase = '') {
-    const promises = this[p.parsers]
-      .map((parser) => parser.parse(phrase));
+  parse(text = '') {
+    text = String(text);
+
+    const result = {
+      intent: null,
+    };
+
+    return new Promise((resolve, reject) => {
+      Promise.resolve()
+      // Cleaning phase.
+        .then(() => {
+          return this[p.normaliseCleaner].clean(text);
+        })
+        .then((text) => {
+          result.text = text; // Let's keep the normalised version from now.
+          return this[p.contractionsCleaner].clean(text);
+        })
+        .then((text) => {
+          return this[p.salutationsCleaner].clean(text);
+        })
+        .then((text) => {
+          return this[p.politenessCleaner].clean(text);
+        })
+        .then((cleaned) => {
+          result.cleaned = cleaned;
+          return result;
+        })
+
+        // Parsing phase.
+        .then(() => {
+          return this[p.timeParser].parse(result);
+        })
+        .then((time) => {
+          result.time = time;
+          return this[p.usersParser].parse(result);
+        })
+        .then((recipients) => {
+          result.recipients = recipients;
+          return this[p.actionParser].parse(result);
+        })
+        .then((action) => {
+          result.action = action;
+          return result;
+        })
+
+        // Refining phase.
+        .then((parsedResult) => {
+          return this[p.reminderRefiner].refine(parsedResult);
+        })
+        .then((parsedResult) => {
+          return this[p.queryRefiner].refine(parsedResult);
+        })
+
+        // Final result.
+        .then((parsedResult) => {
+          if (parsedResult.intent === null) {
+            console.error('The intent could not be parsed:', parsedResult);
+            return reject(null);
+          }
 
-    // Return the first non null parsed result.
-    return Promise.all(promises)
-      .then((results) => results.find((result) => !!result))
-      .then((result) => result || null);
+          return resolve(parsedResult);
+        });
```

Comment:
I'm not super happy with all this code but couldn't think of anything better. Any suggestions?